src/java.base/share/classes/sun/security/util/DerInputStream.java

```diff
@@ -661,4 +684,118 @@ static int getDefiniteLength(InputStream in) throws IOException {
+     */
+    public void atEnd() throws IOException {
+        if (available() != 0) {
+            throw new IOException("Extra unused bytes");
+        }
+    }
+
+    /**
+     * Detect if the tag of the next DerValue in the stream matches the rule.
+     *
+     * Attention: tag is an integer casted from a byte. i.e. could be negative.
+     *
+     * @param rule the rule to check for the tag.
+     * @return true if matches, false otherwise or stream is at end.
+     * @throws IOException if an I/O error happens
+     */
+    public boolean seeOptional(Predicate<Integer> rule) throws IOException {
+        return available() > 0 && rule.test(peekByte());
+    }
+
+    /**
+     * Detect if the tag of the next DerValue in the stream is the specified.
+     *
+     * @param tag the expected tag
+     * @return true if is, false otherwise or stream is at end.
+     * @throws IOException if an I/O error happens
+     */
+    public boolean seeOptional(byte tag) throws IOException {
+        return seeOptional(t -> t == (tag & 0xff));
+    }
+
+    /**
+     * Returns the inner DerValue if the next DerValue in the stream is
+     * an EXPLICIT context-specific value tagged by {@code n}.
+     *
+     * @param n the expected tag
+     * @return the inner DerValue, or empty if not found or stream at end
+     * @throws IOException if an I/O error happens
+     */
+    public Optional<DerValue> getOptionalExplicitContextSpecific(int n)
+            throws IOException {
+        if (seeOptionalContextSpecific(n)) {
+            DerValue v = getDerValue(); // [n]
+            DerValue sub = v.data.getDerValue(); // inside [n]
+            v.data.atEnd(); // make sure there is only one inner value
+            return Optional.of(sub);
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Returns the restored DerValue if the next DerValue in the stream is
+     * an IMPLICIT context-specific value tagged by {@code n}.
+     *
+     * @param n the expected tag
+     * @param tag the real tag for the IMPLICIT type
+     * @return the restored DerValue, or empty if not found or stream at end
+     * @throws IOException if an I/O error happens
+     */
+    public Optional<DerValue> getOptionalImplicitContextSpecific(int n, byte tag)
+            throws IOException {
+        if (seeOptionalContextSpecific(n)) {
+            DerValue v = getDerValue(); // [n]
+            v.resetTag(tag); // restore tag because IMPLICIT has overwritten it
+            return Optional.of(v);
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Detect if the next DerValue in the stream is a context-specific value
+     * tagged by {@code n}.
+     *
+     * @param n the expected tag
+     * @return true if is, false otherwise or stream is at end.
+     * @throws IOException if an I/O error happens
+     */
+    public boolean seeOptionalContextSpecific(int n) throws IOException {
```

Comment:
Consider moving this seeOptionalXXX upward so that all seeXXX methods are together for easier reference.

Reply:
Good suggestion.